Addition of a utility class for cycling writes

Titantompa opened the pull request:
Hey,

I wrote a utility class for your library and though I'd ask you if you'd like to include it in your main library (rather than me registering yet another minimal library that can only be used with your's).

My use case is that I have various devices that need the ability to retain their state between power cycles. While they may not lose power that frequently the changes to their state are much more frequent. I also expect them to operate indefinitely without any periodic mentainance. So I figured I can extend the expected use of an eeprom by simply spreading the writes evenly over the available pages - hoping that it will simply suffice.

So, to do this I implemented this template class that writes the data to the next higher position in memory each time and starts over at address zero when the end of the eeprom has been reached. In order to know what was the last writted position it adds a header to the data containing an incremental version number. At initialization it locates the highest version number (by a partitioning search) and simply continues where it left off.

So, please have a look at it and add it to your lib if you think it is suitable! (I've tried to keep the code similar to how you have written yours, so that it wouldn't look haphazard.) Please squash the commit so that all my interim ones are left out!

Oh, and I also updated the code that figures out the page size, since it didn't handle larger eeproms.

Cheers

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,239 @@
+#pragma once
+#ifndef CYCLICACCESS_H
```
If you use **pragma once** the construction with **ifndef** is not needed

Titantompa replied:
I read somewhere that it's adviseable to use both, since some compilers still don't support the **pragma once** directive. Not sure to what extent it is necessary so I'll remove it for consistency with your code.

RobTillaart replied:
Valid point, 
however most compilers do support #pragma as I did not get any issue in any of my libs about it. 
Furthermore, as your class depends on mine and mine has not both it made less sense.

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,239 @@
+            memcpy(tmp, &_currentVersion, sizeof(_currentVersion));
+            memcpy(tmp+sizeof(_currentVersion), buffer, sizeof(T));
+
+            Serial.println("Writing "+String(buffer_length)+" bytes to eeprom");
```
Debugging statements should not be this explicit as it may corrupt the output of the user of the class.
==> use a **#ifdef debug**  construct ?

Better not use the String Class as it might only be included for this message.
2 or 3 (conditional) Serial.print statements would be better

Titantompa replied:
Oh, I am really sorry for what turned out to be a bit of a sloppy pull request. :-/ Those shouldn't have been left in there, I thought I had scrubbed it clean of the debug stuff.

RobTillaart replied:
No need to apologize, these things happen and that is where reviews are for.

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,239 @@
+        }   
+
+    private:
+        uint16_t _pageSize;
```
Do all these variables need to be 16 bit - did not check them in detail.
Saving a few bytes is always good for the smallest of the microprocessors.

Titantompa replied:
I had a look at it and the _bufferSize turned out to be redundant.
The _pageSize could be shrunk to 8 bits, as I've only seen 256 byte pages so far and the main code limits it to 128.
But I think the others are as they need to be - even though I get a nagging thought about introducing something that makes it easier for future users to detect if they are trying to use an eeprom that would exceeds the current limits...

RobTillaart replied:
OK, you gave them a serious second thought, and when needed they can still be updated.

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,239 @@
+        }
+};
+
+#endif // CYCLICACCESS_H
```
See **pragma once**

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,239 @@
+            if(_isEmpty)
+                return false;
+
+            Serial.println("Reading "+String(sizeof(T))+" bytes from eeprom");
```
See other remark about printing

RobTillaart commented:
Question:
If as system restarts, how does it know the last slot used to read the last good value from EEPROM?
Reuse same location is straightforward, however your class explicitly use multiple locations to reduce 'overused locations'

RobTillaart commented:
Please also check the CI build .... als it need to be green before merge 
- https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md

RobTillaart commented:
Q: should the class have some statistics? 
(No is a valid answer...)

Titantompa commented:
> Question:
> If as system restarts, how does it know the last slot used to read the last good value from EEPROM?
> Reuse same location is straightforward, however your class explicitly use multiple locations to reduce 'overused locations'



> Question:
> If as system restarts, how does it know the last slot used to read the last good value from EEPROM?
> Reuse same location is straightforward, however your class explicitly use multiple locations to reduce 'overused locations'

Since it does the writes in sequential order from the start of the general idea is that you, starting from the first slot, simply fast forward as long as the version number is higher than the previous one. This is not very optimal, if you have hundreds or thousands of pages, so I wrote a simple partitioning search.

It checks the centre-point of the search set - if the version number is higher than the start it means that the last must be after the centrepoint so it reduces the search set to the upper half and "recurses" into that. Similarly, if the centre-point has lower than that of the first slot the highest version number must be in the lower half of the search set...
I'm not sure how to express the number of comarisons mathematically, but it does the same number of comparisons regardless of where the highest version is. E g if you have 512 pages it will find the highest version number in nine iterations. 

This is implemented in the `initialize()` method.

Titantompa commented:
> Q: should the class have some statistics?
> (No is a valid answer...)

I had a few thoughts but at the same time felt like they were more of a debugging or anecdotal nature?

Adding a few more fields to the header would allow for a lot more interesting metrics but I'm hesitant to using eeprom memory as it could significantly reduce the expected number of writes to the chip (if they'd cause the slot to require one more write page).

Maybe if it could provide a metric that indicates how many of the expected writes have been consumed? E g:
u = x / ( v / (p / b) )
u = used % of expected writes
x = expected available writes
p = total number of pages
b = buffer page size 

Maybe there is a reasonable default value to use for the expected number of writes? So that it could be written as a simple method that returns a float between 0.0 and 1.0?

RobTillaart commented:
> > Question:
> > If as system restarts, how does it know the last slot used to read the last good value from EEPROM?
> > Reuse same location is straightforward, however your class explicitly use multiple locations to reduce 'overused locations'
> 
> Since it does the writes in sequential order from the start of the general idea is that you, starting from the first slot, simply fast forward as long as the version number is higher than the previous one. This is not very optimal, if you have hundreds or thousands of pages, so I wrote a simple partitioning search.
> 
> It checks the centre-point of the search set - if the version number is higher than the start it means that the last must be after the centrepoint so it reduces the search set to the upper half and "recurses" into that. Similarly, if the centre-point has lower than that of the first slot the highest version number must be in the lower half of the search set...
> I'm not sure how to express the number of comarisons mathematically, but it does the same number of comparisons regardless of where the highest version is. E g if you have 512 pages it will find the highest version number in nine iterations.
> 
> This is implemented in the `initialize()` method.

This strategy is called binary search and the number of steps needed are O( 2log(size) ). 
That means that if there are 64 elements you need  O(2log(64)) steps = 6 iterations, 
100 elements need 6.64... = 7 iterations, 1024 elements need ~10 iterations, a million = 20 steps etc

RobTillaart commented:
A review of initialize shows some (minor) improvements.
```cpp
bool initialize()
{
  uint16_t startSlot, probeSlot, endSlot;
  uint32_t current;

  _eeprom->readBlock(0,  (uint8_t *) &current, sizeof(uint32_t));

  // Memory is blank == first time use
  if (current == ~0UL)
  {
    _isEmpty = true;
    _currentSlot = 0;
    _currentVersion = 0;
    _isInitialized = true;
    return true;
  }
  
  // Find the current block by binary search
  startSlot = 0;
  endSlot   = (_totalPages/_bufferPages) - 1;         // Index of last slot
  probeSlot = startSlot + ((endSlot - startSlot)/2);  // Midway between start and end

  while (startSlot < probeSlot)
  {
    _eeprom->readBlock(probeSlot * _pageSize * _bufferPages, (uint8_t *) &current, sizeof(uint32_t));
    if ( current == ~0UL)
    {
      endSlot   = probeSlot;
    }
    else
    {
      startSlot = probeSlot;
    }
    probeSlot = startSlot + ((endSlot - startSlot)/2);
  }
  _currentSlot = startSlot;
  _currentVersion = current;
  _isEmpty = false;
  _isInitialized = true;
  return true;
};
```

Some minor thoughts 
- _pageSize * _bufferPages, could be out of the loop as it is a constant.
- initialize always returns true, which is pointless at the moment but can be a preparation for error indication. Better to return an uint8_t as that can have 256 distinct values from which e.g. 0 == OK

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,231 @@
+         * This must be done if the eeprom already contains data or if the
+	 * structure of the data stored changes.
+         */
+        void format()
```
Might be faster when using 
```cpp
void format()
{
  _eeprom->setBlock(0, 0xFF, _totalPages * _pageSize);  // this is effectively eeprom size...
  _isEmpty = true;
  _currentSlot = 0;
  _isInitialized = true;
}
```
 you could do a performance test, setBlock uses a larger buffer so it will do less steps.

Thinking about it
 a **format(uint8_t)** function should be in the EEPROM class itself as it just wipes the whole EE.

your format would then become
```cpp
void format()
{
  _eeprom->format(0xFF);
  _isEmpty = true;
  _currentSlot = 0;
  _isInitialized = true;
}
```

Titantompa replied:
I agree that it would be neat to erase the entire eeprom, but from what I gather there is no shortcut (e g erase page command) so we'd have to write 0xFF to the entire chip, right? While it internally, in the eeprom, wouldn't make a difference, my approach reduces the I2C traffic to a minimum. Especially when the buffer spans multiple pages.

But this is more or less speculation... 😄 I will make a test.

Titantompa replied:
From a few timing runs with a buffer size that is less than the page size, and with 512 write pages, it takes the old implementation ~1.5 seconds to do a format compared to ~8 seconds using the setBlock() call.

Titantompa commented:
> A review of initialize shows some (minor) improvements.
> 
> ```c++
> bool initialize()
> {
>   uint16_t startSlot, probeSlot, endSlot;
>   uint32_t current;
> 
>   _eeprom->readBlock(0,  (uint8_t *) &current, sizeof(uint32_t));
> 
>   // Memory is blank == first time use
>   if (current == ~0UL)
>   {
>     _isEmpty = true;
>     _currentSlot = 0;
>     _currentVersion = 0;
>     _isInitialized = true;
>     return true;
>   }
>   
>   // Find the current block by binary search
>   startSlot = 0;
>   endSlot   = (_totalPages/_bufferPages) - 1;         // Index of last slot
>   probeSlot = startSlot + ((endSlot - startSlot)/2);  // Midway between start and end
> 
>   while (startSlot < probeSlot)
>   {
>     _eeprom->readBlock(probeSlot * _pageSize * _bufferPages, (uint8_t *) &current, sizeof(uint32_t));
>     if ( current == ~0UL)
>     {
>       endSlot   = probeSlot;
>     }
>     else
>     {
>       startSlot = probeSlot;
>     }
>     probeSlot = startSlot + ((endSlot - startSlot)/2);
>   }
>   _currentSlot = startSlot;
>   _currentVersion = current;
>   _isEmpty = false;
>   _isInitialized = true;
>   return true;
> };
> ```
> 
> Some minor thoughts
> 
> * _pageSize * _bufferPages, could be out of the loop as it is a constant.
> * initialize always returns true, which is pointless at the moment but can be a preparation for error indication. Better to return an uint8_t as that can have 256 distinct values from which e.g. 0 == OK

I've moved the calculation of the buffer length to outside the loop. 👍 

I also made it propagate read errors from the I2C_eeprom class itself so it now actually can return either true or false. As there is no further information from the underlying calls I thought it better to simply return a boolean. If the call to begin() is improved in the future, featuring an int return value carrying more detailed failure information it would be better to remove or make the current interface obsolete - with a reference to the new improved one so that users can adapt their code appropriately.

This resonates a little with the read() and write() calls where the first has no error information and the latter would need to encapsule or translate the Wire error codes to make room for the possibility that the instance hasn't been initialized properly. Facing the complexity of it, I've opted to keep it light weight and a little crude.

Surprisingly this code also featured a re-surfaced bug, with finding the current of an even number of slots, that I fixed. Again.

Titantompa commented:
> > Q: should the class have some statistics?
> > (No is a valid answer...)
> 
> I had a few thoughts but at the same time felt like they were more of a debugging or anecdotal nature?
> 
> Adding a few more fields to the header would allow for a lot more interesting metrics but I'm hesitant to using eeprom memory as it could significantly reduce the expected number of writes to the chip (if they'd cause the slot to require one more write page).
> 
> Maybe if it could provide a metric that indicates how many of the expected writes have been consumed? E g:
> u = x / ( v / (p / b) )
> u = used % of expected writes
> x = expected available writes
> p = total number of pages
> b = buffer page size
> 
> Maybe there is a reasonable default value to use for the expected number of writes? So that it could be written as a simple method that returns a float between 0.0 and 1.0?

I added a getMetrics() method that returns the number of slots and the number of writes - that way you can calculate the average number of page writes to guesstimate how far along you are.

RobTillaart commented:
Did you do a performance test with the proposed format (setBlock variant) ?
As most applications will use it seldom it does not matter that much.

I will review coming week (If I find a slot maybe today)

RobTillaart commented on I2C_eeprom_cyclic_store.h:
```diff
@@ -0,0 +1,262 @@
+      * @param[out] writeCounter The total number of write to the eeprom since the last format (or first use).
+      * @return True if the instance is initialized and the eeprom is not blank.
+      */
+    bool getMetrics(uint16_t &slots, uint32_t &writeCounter)
```
If the EEPROM is empty it still should know its nr of pages ? 
I think only when the EEPROM is not initialized the return of false is correct?
Or miss I some thought?

Titantompa replied:
Hm, yes, I was mistakenly thinking that the writeCounter wouldn't have a value - but it should be zero in case the eeprom is empty!

Titantompa commented:
> Did you do a performance test with the proposed format (setBlock variant) ?
> As most applications will use it seldom it does not matter that much.
> 
> I will review coming week (If I find a slot maybe today)

Yes, the setBlock() variant takes approximately five times longer.

RobTillaart reviewed (commented):
> > Did you do a performance test with the proposed format (setBlock variant) ?
> > As most applications will use it seldom it does not matter that much.
> > I will review coming week (If I find a slot maybe today)
> 
> Yes, the setBlock() variant takes approximately five times longer.

ouch...

RobTillaart commented:
Idea, although quite some work
- would it be better to create a derived class with this extra functionality?
- the user only has one object to consider
- interfaces may be simpler
- would make it easier to move functionality from derived -> base when appropriate..

Titantompa commented:
> Idea, although quite some work
> 
> * would it be better to create a derived class with this extra functionality?
> * the user only has one object to consider
> * interfaces may be simpler
> * would make it easier to move functionality from derived -> base when appropriate..

A valid point, but I'd argue against a specialization as it skyrockets the complexity. There are undisputed advantages of the composition pattern...

It would maybe be better to enable this class to act as a facade by adding a variation of `begin()` that creates an instance of `I2C_eeprom` that is kept behind the scenes? It adds a lot of code (relatively) since this would have to include overloads for all parameter variations to the `I2C_eeprom` constructor and the `begin()` method. I'm also not sure it would make life easier for a user as the number of options would increase to the point where it might obfuscate things..?

One thing that could be considered for future versions is if it would be worthwhile to create abstract base classes for this and `I2C_eeprom` so that mock implementations could be used in unit tests?

RobTillaart commented:
I will squash and merge the commits , keeping the unique commit comments. 
OK?

Titantompa commented:
> I will squash and merge the commits , keeping the unique commit comments.
> OK?

Absolutely! Thanks! 👍

RobTillaart commented:
@Titantompa
done
